actionpack/lib/action_controller/test_case.rb

```diff
@@ -468,6 +468,8 @@ def head(action, **args)
       # Note that the request method is not verified.
       def process(action, method: "GET", params: nil, session: nil, body: nil, flash: {}, format: nil, xhr: false, as: nil)
         check_required_ivars
+        @controller.purge_new_instance_variables_from_previous_requests
+        @controller.record_instance_variables
```

Comment:
Couldn't we instantiate a new controller instead?

Reply:
No, because then you can’t set ivars on the controller before any request runs.

While that’s not a good idea it’s possible and I’m sure lots of tests in the wild do it.

An example use case would be setting the current user.